modules/openthread/platform/diag.c

```diff
@@ -291,25 +319,152 @@ otError otPlatDiagGpioGetMode(uint32_t aGpio, otGpioMode *aMode)
 
 static otError startModCarrier(otInstance *aInstance, uint8_t aArgsLength, char *aArgs[])
 {
-	ARG_UNUSED(aInstance);
-	ARG_UNUSED(aArgsLength);
-
 	bool enable = true;
 	uint8_t data[OT_RADIO_FRAME_MAX_SIZE + 1];
 
 	if (aArgsLength <= 0) {
 		return OT_ERROR_INVALID_ARGS;
 	}
 
+	if (!otPlatDiagModeGet() && (sTransmitMode != diag_transmit_mode_idle ||
+				     sTransmitMode != diag_transmit_mode_modcarrier)) {
+		return OT_ERROR_INVALID_STATE;
+	}
+
 	if (strcmp(aArgs[0], "stop") == 0) {
 		enable = false;
+		sTransmitMode = diag_transmit_mode_idle;
 	} else {
 		if (hex2bin(aArgs[0], strlen(aArgs[0]), data, ARRAY_SIZE(data)) == 0) {
 			return OT_ERROR_INVALID_ARGS;
 		}
+		sTransmitMode = diag_transmit_mode_modcarrier;
 	}
 
 	return platformRadioTransmitModulatedCarrier(aInstance, enable, data);
 }
 
 #endif
+
+void otPlatDiagAlarmCallback(otInstance *aInstance)
+{
+	uint32_t now;
+	otRadioFrame *sTxPacket;
+	const uint16_t diag_packet_len = 30;
+
+	if (sTransmitMode == diag_transmit_mode_packets) {
+		if ((sTxCount > 0) || (sTxCount == -1)) {
+			sTxPacket = otPlatRadioGetTransmitBuffer(aInstance);
+
+			sTxPacket->mInfo.mTxInfo.mTxDelayBaseTime = 0;
+			sTxPacket->mInfo.mTxInfo.mTxDelay = 0;
+			sTxPacket->mInfo.mTxInfo.mMaxCsmaBackoffs = 0;
+			sTxPacket->mInfo.mTxInfo.mMaxFrameRetries = 0;
+			sTxPacket->mInfo.mTxInfo.mRxChannelAfterTxDone = sChannel;
+			sTxPacket->mInfo.mTxInfo.mTxPower = OT_RADIO_POWER_INVALID;
+			sTxPacket->mInfo.mTxInfo.mIsHeaderUpdated = false;
+			sTxPacket->mInfo.mTxInfo.mIsARetx = false;
+			sTxPacket->mInfo.mTxInfo.mCsmaCaEnabled = false;
+			sTxPacket->mInfo.mTxInfo.mCslPresent = false;
+			sTxPacket->mInfo.mTxInfo.mIsSecurityProcessed = false;
+
+			sTxPacket->mLength = diag_packet_len;
+
+			for (uint8_t i = 0; i < diag_packet_len; i++) {
+				sTxPacket->mPsdu[i] = i;
+			}
+
+			otPlatRadioTransmit(aInstance, sTxPacket);
+
+			if (sTxCount != -1) {
+				sTxCount--;
+			}
+
+			now = otPlatAlarmMilliGetNow();
+			otPlatAlarmMilliStartAt(aInstance, now, sTxPeriod);
+		} else {
+			sTransmitMode = diag_transmit_mode_idle;
+			otPlatAlarmMilliStop(aInstance);
+			otPlatLog(OT_LOG_LEVEL_DEBG, OT_LOG_REGION_PLATFORM, "Transmit done");
+		}
+	}
+}
+
+static otError processTransmit(otInstance *aInstance, uint8_t aArgsLength, char *aArgs[])
+{
+	otError error = OT_ERROR_NONE;
+	long value;
+	uint32_t now;
+
+	if (aArgsLength == 0) {
+		diag_output("transmit will send %" PRId32 " diagnostic messages with %" PRIu32
+			    " ms interval\r\n",
+			    sTxRequestedCount, sTxPeriod);
+
+	} else if (strcmp(aArgs[0], "stop") == 0) {
+		if (sTransmitMode == diag_transmit_mode_idle) {
+			return OT_ERROR_INVALID_STATE;
+		}
+
+		otPlatAlarmMilliStop(aInstance);
+		diag_output("diagnostic message transmission is stopped\r\nstatus 0x%02x\r\n",
+			    error);
```

Comment:
```suggestion
		diag_output("diagnostic message transmission is stopped\r\n");
```
`error` is not updated in this case so it is always `OT_ERROR_NONE`